crypto/threads_pthread.c

```diff
@@ -42,12 +44,475 @@
+    return ret;
+}
+
+# define __ATOMIC_OR_FETCH(p, v, o) __fallback_atomic_or_fetch(p, v)
+#endif
+
+static CRYPTO_THREAD_LOCAL rcu_thr_key;
+
+struct rcu_thr_data;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
+ */
+# define READER_SHIFT 0
+# define ID_SHIFT 32
+# define READER_SIZE 16
+# define ID_SIZE 32
+
+# define READER_MASK     (((uint64_t)1 << READER_SIZE)-1)
+# define ID_MASK         (((uint64_t)1 << ID_SIZE)-1)
+# define READER_COUNT(x) (((uint64_t)(x) >> READER_SHIFT) & READER_MASK)
+# define ID_VAL(x)       (((uint64_t)(x) >> ID_SHIFT) & ID_MASK)
+# define VAL_READER      ((uint64_t)1 << READER_SHIFT)
+# define VAL_ID(x)       ((uint64_t)x << ID_SHIFT)
+
+/*
+ * This is the core of an rcu lock
+ * it tracks the the readers and writers
+ * for the current quiescence point for
+ * a given lock
+ * users is the 64 bit value that stores
+ * the READERS/ID as defined
+ * above
+ *
+ */
+struct rcu_qp {
+    uint64_t users;
+};
+
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ * count is the threads current recursion count
+ */
+struct rcu_thr_data {
+    struct rcu_qp *qp;
+    int count;
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint64_t id_ctr;
+    struct rcu_qp *qp_group;
+    size_t group_count;
+    uint64_t reader_idx;
+    uint32_t next_to_retire;
+    uint64_t current_alloc_idx;
+    uint32_t writers_alloced;
+    pthread_mutex_t write_lock;
+    pthread_mutex_t alloc_lock;
+    pthread_cond_t alloc_signal;
+    pthread_mutex_t prior_lock;
+    pthread_cond_t prior_signal;
+};
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data = 
+                        (struct rcu_thr_data *)CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data)
+        CRYPTO_free(data, __FILE__, __LINE__);
+    CRYPTO_THREAD_set_local(&rcu_thr_key, NULL);
+}
+
+static void ossl_rcu_init(void)
+{
+    CRYPTO_THREAD_init_local(&rcu_thr_key, NULL);
+}
+
+/* Read side acquisition of the current qp */
+static struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+{
+    uint64_t qp_idx;
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /* get the current qp index */
+    for(;;) {
+        qp_idx = (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_SEQ_CST);
+        __ATOMIC_ADD_FETCH(&ilock->qp_group[qp_idx].users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+        /* if the idx hasn't changed, we're good, else try again */
+        if (qp_idx == (uint64_t)__ATOMIC_LOAD_N(&ilock->reader_idx, __ATOMIC_SEQ_CST))
```

Comment:
Can you explain, or do you have a reference, of why all those are SEQ_CST? I assumed that acquire would be enough.

Reply:
I don't, it was an artifact of an earlier implementation in which I was potentially incrementing and decrementing in the same function, where it seemed needed, I can reduce the strength of the ordering I think